src/com/xmartlabs/slackbot/manager/WorkTimeReportManager.kt

```diff
@@ -0,0 +1,281 @@
+            }
+            .onEach { report -> sendDMessageToUser(report.slackUser, report.togglReport, from, to) }
+            .also { entries -> sendReportSummaryMessage(entries, now, from, to) }
+    }
+
+    private suspend inline fun <reified ToggleReport : TogglUserEntryReport> generateWorkTimeReport(
+        fromDate: LocalDate,
+        toDate: LocalDate,
+    ): List<UserWorkTimeReport<ToggleReport>> = coroutineScope {
+        val start = LocalDateTime.now()
+        val bambooReportDeferred =
+            async {
+                getBambooReport(fromDate, toDate)
+                    .associateBy { it.bambooUser.workEmail }
+            }
+        val usersDeferred = async {
+            UserSlackRepository.getUsers()
+                .associateBy { it.profile.email }
+        }
+        val togglReportDeferred: Deferred<List<ToggleReport>> = async {
+            @Suppress("UNCHECKED_CAST")
+            when (ToggleReport::class.java) {
+                FullTogglUserEntryReport::class.java -> TogglReportRepository.generateFullReport(
+                    fromDate.atTime(LocalTime.MIN), toDate.atTime(LocalTime.MAX)
+                )
+                SimpleTogglUserEntryReport::class.java -> TogglReportRepository.generateSimpleReport(fromDate, toDate)
+                else -> throw IllegalStateException("type is not valid")
+            } as List<ToggleReport>
+        }
+        val bambooReport = bambooReportDeferred.await()
+        val slackUsers = usersDeferred.await()
+        togglReportDeferred.await()
+            .mapNotNull { togglReport ->
+                val slackUser = slackUsers[togglReport.togglUser.email]
+                if (slackUser == null) {
+                    logger.warn("Slack user not found. Toggl user: ${togglReport.togglUser}")
+                    null
+                } else {
+                    UserWorkTimeReport(
+                        slackUser = slackUser,
+                        togglReport = togglReport,
+                        bambooReport = bambooReport[togglReport.togglUser.email]
+                    )
+                }
+            }
+            .also {
+                val calculationTime = Duration.between(start, LocalDateTime.now())
+                    .toPrettyString(true)
+                val from = fromDate.format(MessageManager.LOCAL_DATE_FORMATTER)
+                val to = toDate.format(MessageManager.LOCAL_DATE_FORMATTER)
+                logger.info("Work hours report from $from to $to generated in $calculationTime)")
+            }
+    }
+
+    private suspend fun getBambooReport(from: LocalDate, to: LocalDate): List<BambooWorkTimeReport> = coroutineScope {
+        require(!from.isAfter(to))
+        (async { BambooHrReportsRemoteSource.getUsers() } to
+                async { BambooHrReportsRemoteSource.getTimeOff(from, to) })
+            .let { it.first.await() to it.second.await() }
+            .let { (users, timeOffs) ->
```

Comment:
What about something like, and removing the let? 

```suggestion
        val (users, timeOffs) = (async { BambooHrReportsRemoteSource.getUsers() }.await() to
                async { BambooHrReportsRemoteSource.getTimeOff(from, to).await() })
```

Reply:
and you map the generated collection?

Reply:
It's not the same right, in that way getUsers will be executed before getTimeOff right?

Reply:
Yes I tested, 
```
    GlobalScope.launch {
        val start = Date()
        val (value1, value2) = async { delay(1000); 1 }.await() to async { delay(1000); 1 }.await()
        println("*++++++++++" + (Date().time - start.time))
    }
```
the output is 2010

However if I use: 
```
    GlobalScope.launch {
        val start = Date()
        val (value1, value2) = (async { delay(1000); 1 } to async { delay(1000); 1 })
            .let { it.first.await() to it.second.await() }
        println("*++++++++++" + (Date().time - start.time))
    }
```
The output is 1008

Reply:
what about using variables? instead f let?